app/src/main/java/org/mozilla/tv/firefox/ext/Js.kt

```diff
@@ -17,12 +19,79 @@ object Js {
 
     const val PAUSE_VIDEO = "document.querySelectorAll('video').forEach(v => v.pause());"
 
+    /**
+     * Due to a bug in the Fire OS WebView (#2540, #2541), fullscreen mp4 videos will appear partially off screen if the
+     * page, before entering fullscreen, is at a scroll position greater than 0. We can correct these videos by caching
+     * the scroll position before the page was fullscreened and restoring this scroll position after entering fullscreen
+     * mode.
+     *
+     * This fix does not work for all sites. In particular, it does not help YouTube desktop.
+     *
+     * This fix has a side effect of sometimes changing the page scroll position after the user leaves fullscreen.
+     *
+     * webm videos were working fine before this fix and this fix does not appear to impact them; I have not tried other
+     * video types types because I don't know of any other supported ones.
+     */
+    object MP4TranslationWorkaround {
+        private const val CACHED_SCROLL_POSITION = "_firefoxTV_cachedScrollPosition"
+
+        /**
+         * Caches the page scroll position before an element is fullscreened via a scroll position observer. Ideally,
+         * we'd cache the scroll position when fullscreen is pressed but there are no appropriate hooks for that
+         * ([Session.Observer.onFullScreenChanged] is too late) so we use a global scroll observer instead.
+         */
+        val OBSERVE_SCROLL_POSITION = """
+var $CACHED_SCROLL_POSITION;
+var _firefoxTV_lastTimeout;
+var _firefoxTV_isScrollPositionObserverLoaded;
+(function () {
+    /* Ensure we only add one listener. */
+    if (_firefoxTV_isScrollPositionObserverLoaded) { return; }
+    _firefoxTV_isScrollPositionObserverLoaded = true;
+
+    console.log('adding scroll event listener');
+    window.addEventListener('scroll', (e) => {
+        /* During the transition to fullscreen, the WebView may or may not change the scroll position to an incorrect
+         * value so we add a short delay before caching. This creates a race condition & is imperfect: on a slow page,
+         * it's possible the incorrect value will cache before we restore the fullscreen value so we'll restore the
+         * incorrect position. Also, if the user is fast enough, they can open fullscreen before we cache the value.
+         * This value was set through manual testing. I did not find better solutions. */
+        if (typeof _firefoxTV_lastTimeout !== 'undefined') {
```

Comment:
Why do we need to clear the timeout here but not in `UPDATE_FULLSCREEN_SCROLL_POSITION`?

Reply:
It's an optimization: we could allow the timeout to fire for every scroll event and the fix would still work correctly but given how fast scroll events fire, there could be a non-trivial impact on the JS thread event queue.

I'll add a comment.